test/distributed/_composable/fsdp/test_fully_shard_init.py

```diff
@@ -993,5 +993,105 @@ def test_hsdp_broadcast_across_replicas(self):
         model(inp).sum().backward()
 
 
+class TestFullyShardShardPlacementFn(FSDPTestMultiThread):
+    @property
+    def world_size(self) -> int:
+        return 4
+
+    def _init_models(self):
+        torch.manual_seed(42)
+        model_args = ModelArgs(n_layers=3, dropout_p=0.0)
+        model = Transformer(model_args)
+        for param in model.parameters():
+            dist.broadcast(param.detach(), src=0)
+        ref_model = copy.deepcopy(model)
+        return model, ref_model
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_init_1d_shard_placement_fn_shard_largest_dim(self):
+        model, ref_model = self._init_models()
+
+        def shard_placement_fn(param: nn.Parameter) -> Optional[Shard]:
+            largest_dim = -1
+            largest_dim_size = -1
+            for dim, dim_size in enumerate(param.shape):
+                if dim_size > largest_dim_size:
+                    largest_dim = dim
+                    largest_dim_size = dim_size
+            assert largest_dim >= 0, f"{param.shape}"
+            return Shard(largest_dim)
+
+        for layer in model.layers:
+            fully_shard(layer, shard_placement_fn=shard_placement_fn)
+        fully_shard(model, shard_placement_fn=shard_placement_fn)
+
+        for param, ref_param in zip(model.parameters(), ref_model.parameters()):
+            full_param = param.full_tensor()
```

Comment:
if I print params here, some params would be shard(1), and some params are shard(0)? 

We expect state dict to be composable with this PR because DTensor full_tensor/redistribute will handle placement = shard(0) / shard(1) gracefully?

Reply:
> if I print params here, some params would be shard(1), and some params are shard(0)?

Yep!

> We expect state dict to be composable with this PR because DTensor full_tensor/redistribute will handle placement = shard(0) / shard(1) gracefully?

That is my understanding.